### PR TITLE
added flat function

### PR DESCRIPTION
**Pull request by sithmel**

### CODE REVIEW

**Comment by KSXGitHub** [comment]
As you said in https://github.com/sithmel/iter-tools/issues/75#issuecomment-444634511, flatten infinitely seems impossible, but what about `flatten(fn, iterable)`?

File: src/flat.mjs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import ensureIterable from './internal/ensure-iterable'
+
+function * flat (depth, iterable) {
+  if (depth === 0) {
+    yield * ensureIterable(iterable)
+  } else {
+    for (const iter of ensureIterable(iterable)) {
+      if (iter[Symbol.iterator]) {
```
Comment (KSXGitHub) [inline src/flat.mjs]: Should it be `if (typeof iter[Symbol.iterator] === 'function')`?

Comment (sithmel) [reply src/flat.mjs]: This is based on the assumption that if you implement ```iter[Symbol.iterator]``` with anything different from a function or undefined, you are a bad person and deserve to spend an afternoon debugging. But I am going to fix it for sake of readability :smile:

Comment (conartist6) [reply src/flat.mjs]: The whole check looks superfluous to me, it's already done in `ensureIterable` so we will always end up in the first branch here.

Comment (sithmel) [reply src/flat.mjs]: I am not sure I follow.
ensureIterable ensures that the original iterable is an iterable. Then I check item by item if they are iterable or not.

Comment (conartist6) [reply src/flat.mjs]: I'm sorry, I was reading wrong. The code is right.

Comment (sithmel) [reply src/flat.mjs]: Better safe than sorry. Thanks for checking ;-)

**Comment by sithmel** [comment]
@KSXGitHub can you please have a look at the TS signature? you can now use a function instead of a depth

File: src/flat.mjs [hunk 2]
```diff
@@ -1,29 +1,39 @@
...
       }
+    } else {
+      yield iterable
     }
   }
+  return _flat(0, ensureIterable(iterable))
 }
 
-export default function curriedFlat (depth, iterable) {
+export default function curriedFlat () {
```
Comment (KSXGitHub) [inline src/flat.mjs]: I see you're using `arguments` variable which was never recommended in JavaScript.
A better alternative would be rest parameters.

```javascript
export default function curriedFlat (...args)
```

**Comment by KSXGitHub** [comment]
@sithmel 

> @KSXGitHub can you please have a look at the TS signature? you can now use a function instead of a depth

I had a look at that, and it isn't correct.

BTW, I already created https://github.com/sithmel/iter-tools/pull/81 for `flat(iterable)` and `flat(depth, iterable)`. I will correct your type definition of `flat(fn, iterable)` when this pull is merged (so that I don't have to change my type definition every time you add changes to the code)

**Comment by KSXGitHub** [comment]
@conartist6 I would prefer string not to be flatten

File: src/flat.mjs (hunk 2 above)
Comment (sithmel) [reply src/flat.mjs]: Rest args are nicer and more readable. Ok

**Comment by KSXGitHub** [comment]
When will you guys merge this? I have a project that uses `R.flatten()` right now and I would be happy to switch to this.

**Comment by sithmel** [comment]
@KSXGitHub I will merge tomorrow, unless there is some objection

File: src/flat.mjs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  if (typeof depth === 'number') {
+    return (currentDepth, iter) =>
+      currentDepth <= depth && typeof iter[Symbol.iterator] === 'function' && typeof iter !== 'string'
+  }
+  throw new Error('flat: "depth" can be a function or a number')
+}
+
+function flat (shouldIFlat, iterable) {
+  function * _flat (currentDepth, iterable) {
+    if (shouldIFlat(currentDepth, iterable)) {
```
Comment (KSXGitHub) [inline src/flat.mjs]: I think you should remove iterable check [here](https://github.com/sithmel/iter-tools/blob/0b5d86e12d1950bac9599bd59a9c83992481e834/src/flat.mjs#L9) and move to this line, right before calling `shouldIFlat()` (`typeof iter[Symbol.iterator] === 'function' && shouldIFlat()`).

The current implementation burdens user to check if `iter` parameter is actually iterable (if the user doesn't, `flat()` will attempt to iterate on non-iterables causing errors).

**Pull request merged**